src/main/java/com/minecolonies/core/entity/ai/workers/service/EntityAIWorkDeliveryman.java

```diff
@@ -241,20 +243,49 @@ private boolean pickupFromBuilding(@NotNull final IBuilding building)
             return false;
         }
 
-        if (ItemStackUtils.isEmpty(handler.getStackInSlot(currentSlot)))
+        if (request.getRequest().getPickupFilter() != null)
         {
-            return false;
+            final ListIterator<ItemStack> iterator = request.getRequest().getPickupFilter().listIterator();
+            while (iterator.hasNext())
+            {
+                final ItemStack filter = iterator.next();
+                if (ItemStackUtils.compareItemStacksIgnoreStackSize(stack, filter))
+                {
+                    final ItemStack activeStack = handler.extractItem(currentSlot, filter.getCount(), false);
+                    transferItemForPickup(activeStack);
+                    if (filter.getCount() == activeStack.getCount())
+                    {
```

Comment:
Feels weird to be iterator based. Can't we just make this a if pickupList.contains(new ItemStorage(stackInCurrentSlot)?

Reply:
The problem occurs when the items are split across several inventory slots, there's a low chance that the items are held in one and the same inventory slot. This one keeps iterating the slots until the request is completely fulfilled.